rest/blip_client_test.go

```diff
@@ -87,9 +82,97 @@ type BlipTestClientRunner struct {
 	SkipSubtest                 map[string]bool // map of sub tests on the blip tester runner to skip
 }
 
-type BodyMessagePair struct {
-	body    []byte
-	message *blip.Message
+type BlipTesterDoc struct {
+	revMode           blipTesterRevMode
+	body              []byte
+	revMessageHistory map[string]*blip.Message // History of rev messages received for this document, indexed by revID
+	revHistory        []string                 // ordered history of revTreeIDs (newest first), populated when mode = revtree
+	HLV               db.HybridLogicalVector   // HLV, populated when mode = HLV
+}
+
+const (
+	revModeRevTree blipTesterRevMode = iota
+	revModeHLV
+)
+
+type blipTesterRevMode uint32
+
+func (doc *BlipTesterDoc) isRevKnown(revID string) bool {
+	if doc.revMode == revModeHLV {
+		version := VersionFromRevID(revID)
+		return doc.HLV.IsVersionKnown(version)
+	} else {
+		for _, revTreeID := range doc.revHistory {
+			if revTreeID == revID {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func (doc *BlipTesterDoc) makeRevHistoryForChangesResponse() []string {
+	if doc.revMode == revModeHLV {
+		// For HLV, a changes response only needs to send cv, since rev message will always send full HLV
+		return []string{doc.HLV.GetCurrentVersionString()}
+	} else {
+		var revList []string
+		if len(doc.revHistory) < 20 {
+			revList = doc.revHistory
+		} else {
+			revList = doc.revHistory[0:19]
+		}
+		return revList
+	}
+}
+
+func (doc *BlipTesterDoc) getCurrentRevID() string {
+	if doc.revMode == revModeHLV {
+		return doc.HLV.GetCurrentVersionString()
+	} else {
+		if len(doc.revHistory) == 0 {
+			return ""
+		}
+		return doc.revHistory[0]
+	}
+}
+
+func (doc *BlipTesterDoc) addRevision(revID string, body []byte, message *blip.Message) {
+	doc.revMessageHistory[revID] = message
+	doc.body = body
+	if doc.revMode == revModeHLV {
+		_ = doc.HLV.AddVersion(VersionFromRevID(revID))
+		doc.body = body
+	} else {
+		// prepend revID to revHistory
+		doc.revHistory = append(doc.revHistory, "")
+		copy(doc.revHistory[1:], doc.revHistory)
+		doc.revHistory[0] = revID
```

Comment:
Think this is equivalent ("Push Front/Unshift" in [SliceTricks](https://go.dev/wiki/SliceTricks))

```suggestion
		doc.revHistory = append([]string{revID}, doc.revHistory...)
```